### PR TITLE
Supporting QC and QC/SSCD

### PR DESCRIPTION
**Pull request by Razumain**
Resolving Issue #63

Now supporting multiple certificate types and
Added support to issue QC and QC/SSCD certificates

### CODE REVIEW

File: keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractCaEngineKeyAndCertificateHandler.java [hunk 1]
```diff
@@ -106,6 +111,20 @@ protected List<X509Certificate> issueSigningCertificateChain(
       final AbstractCertificateModelBuilder<? extends AbstractCertificateModelBuilder<?>> certificateModelBuilder =
           this.createCertificateModelBuilder(signingKeyPair.getPublicKey(), certificateNameModel);
 
+      // Add QC declarations
+      SigningCertificateRequirements certificateRequirements = signRequest.getSigningCertificateRequirements();
+      checkCertificateType(certificateRequirements);
```
Comment (martin-lindstrom) [inline keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractCaEngineKeyAndCertificateHandler.java:116]: This should already have been done. Remove.

File: keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractKeyAndCertificateHandler.java [hunk 2]
```diff
@@ -144,12 +144,7 @@ public void checkRequirements(@Nonnull final SignRequestMessage signRequest,
         .orElseThrow(() -> new InvalidRequestException("Missing certificate requirements"));
 
     // Check certificate type.
-    if (certificateRequirements.getCertificateType() != null) {
-      if (!certificateRequirements.getCertificateType().equals(this.getCaCertificateType())) {
-        throw new InvalidRequestException("Handler does not support certificate type " +
-            certificateRequirements.getCertificateType());
-      }
-    }
+    checkCertificateType(certificateRequirements);
```
Comment (martin-lindstrom) [inline keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractKeyAndCertificateHandler.java:147]: Prefix with this.

**Review by martin-lindstrom — CHANGES_REQUESTED** [review]
You also need to update application.yml in the demo app to reflect the changes.

**Review by martin-lindstrom — APPROVED** [review]

File: keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractCaEngineKeyAndCertificateHandler.java [hunk 3]
```diff
@@ -134,6 +153,9 @@ protected List<X509Certificate> issueSigningCertificateChain(
     catch (final CertificateIssuanceException e) {
       throw new CertificateException("Failed to issue certificate - " + e.getMessage(), e);
     }
+    catch (InvalidRequestException e) {
+      throw new CertificateException("Unsupported certificate type");
```
[the remark this answers is not in the document]
Comment (Razumain) [reply keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractCaEngineKeyAndCertificateHandler.java]: Fixed

File: keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractCaEngineKeyAndCertificateHandler.java [hunk 4]
```diff
@@ -106,6 +111,20 @@ protected List<X509Certificate> issueSigningCertificateChain(
       final AbstractCertificateModelBuilder<? extends AbstractCertificateModelBuilder<?>> certificateModelBuilder =
           this.createCertificateModelBuilder(signingKeyPair.getPublicKey(), certificateNameModel);
 
+      // Add QC declarations
+      SigningCertificateRequirements certificateRequirements = signRequest.getSigningCertificateRequirements();
```
[the remark this answers is not in the document]
Comment (Razumain) [reply keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractCaEngineKeyAndCertificateHandler.java]: fixed

File: keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractCaEngineKeyAndCertificateHandler.java (hunk 1 above)
Comment (Razumain) [reply keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractCaEngineKeyAndCertificateHandler.java]: We agreed to keep the check in to ensure that this function can't be called in violation with configuration constraints.

File: keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractKeyAndCertificateHandler.java [hunk 5]
```diff
@@ -170,6 +165,21 @@ public void checkRequirements(@Nonnull final SignRequestMessage signRequest,
     log.debug("Key and certificate issuing requirements on SignRequest passed");
   }
 
+  /**
+   * Checks that the requested certificate type is consistent with supported certificate types
+   * @param certificateRequirements certificate requirements from the certificate request
+   * @throws InvalidRequestException illegal certificate type requirements
+   */
+  protected void checkCertificateType(final @Nonnull SigningCertificateRequirements certificateRequirements)
```
[the remark this answers is not in the document]
Comment (Razumain) [reply keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractKeyAndCertificateHandler.java]: fixed

File: keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractKeyAndCertificateHandler.java (hunk 2 above)
Comment (Razumain) [reply keycert/base/src/main/java/se/swedenconnect/signservice/certificate/base/AbstractKeyAndCertificateHandler.java]: fixed

**Pull request merged**